### PR TITLE
Feature/timestamp check

### PR DESCRIPTION
**Pull request by iamyulong**
## Summary

Add timestamp check

### CODE REVIEW

**Review by dhedey — CHANGES_REQUESTED** [review]
Nice! But there's a couple of tweaks we need I think

File: radix-engine/src/system/system_callback.rs [hunk 1]
```diff
@@ -1606,6 +1606,46 @@ impl<V: SystemCallbackObject> KernelTransactionExecutor for System<V> {
...
+            let current_time = Instant::new(
+                substate
+                    .into_payload()
+                    .fully_update_and_into_latest_version()
+                    .epoch_milli
+                    / 1000,
+            );
+            if !range.contains(current_time) {
+                return Err(Self::create_rejection_receipt(
+                    RejectionReason::InvalidTimestampRange {
```
[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-engine/src/system/system_callback.rs:1628]: Will do.

File: radix-engine/src/system/system_callback.rs [hunk 2]
```diff
@@ -1606,6 +1606,46 @@ impl<V: SystemCallbackObject> KernelTransactionExecutor for System<V> {
             }
         }
 
+        if let Some(range) = executable.overall_proposer_timestamp_range() {
```
[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-engine/src/system/system_callback.rs:1609]: See `pub fn overall_proposer_timestamp_range(&self) -> Option<&ProposerTimestampRange>`.

Am I missing something?

File: radix-engine-tests/tests/kernel/transaction_executor.rs [hunk 3]
```diff
@@ -45,6 +44,56 @@ fn transaction_executed_before_valid_returns_that_rejection_reason() {
     );
 }
 
+#[test]
```
[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-engine-tests/tests/kernel/transaction_executor.rs:47]: The second case is covered by all the existing "scenarios".

Not sure how much extra value I get from the first case, but will add anyway.

File: radix-engine-tests/tests/kernel/transaction_executor.rs [hunk 4]
```diff
@@ -45,6 +44,56 @@ fn transaction_executed_before_valid_returns_that_rejection_reason() {
...
+    assert_eq!(
+        rejection_error,
+        &RejectionReason::InvalidTimestampRange {
+            range: ProposerTimestampRange {
+                start_timestamp_inclusive: Some(Instant::new(proposer_timestamp_ms / 1000 + 1)),
+                end_timestamp_exclusive: None,
+            },
+            current_time: Instant::new(proposer_timestamp_ms / 1000)
+        }
+    );
```
[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-engine-tests/tests/kernel/transaction_executor.rs:94]: It's covered by all the receipt assertions, see PR diff.

File: radix-transaction-scenarios/generated-examples/cuttlefish/basic_subintents/costings/001--create-accounts.txt [hunk 5]
```diff
@@ -1,14 +1,15 @@
...
 - Tipping Cost (XRD)                                                       ,                        0,      0.0%
 - Royalty Cost (XRD)                                                       ,                        0,      0.0%
-Execution Cost Breakdown                                                   ,                  6757251,    100.0%
+Execution Cost Breakdown                                                   ,                  6797251,    100.0%
 - AfterInvoke                                                              ,                      904,      0.0%
-- AllocateNodeId                                                           ,                     3395,      0.1%
+- AllocateNodeId                                                           ,                     3395,      0.0%
 - BeforeInvoke                                                             ,                     3326,      0.0%
 - CheckReference                                                           ,                    40011,      0.6%
+- CheckTimestamp                                                           ,                    40000,      0.6%
```
[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-transaction-scenarios/generated-examples/cuttlefish/basic_subintents/costings/001--create-accounts.txt:12]: To be clear, are we suggesting that we skip validation if the range is `(-infinity, infinity)`.

We got a `Some(range)` from the executable, thus the costing above.

File: radix-engine-tests/tests/kernel/transaction_executor.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dhedey) [reply radix-engine-tests/tests/kernel/transaction_executor.rs:47]: Discussed on the call, I think it'd be good to have a test:
> Ensure that a v2 transaction without any timestamp assertion is not billed for reading the timestamp.

File: radix-engine-tests/tests/kernel/transaction_executor.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dhedey) [reply radix-engine-tests/tests/kernel/transaction_executor.rs:94]: True, although the receipt assertions are incorrect for the v2 transactions :)

File: radix-engine-tests/tests/kernel/transaction_executor.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dhedey) [reply radix-engine-tests/tests/kernel/transaction_executor.rs:47]: > Not sure how much value we can get from the first case, but will add anyway.

Well imo it verifies the three cases the node cares about:
* `RejectionReason::TransactionProposerTimestampNotYetValid`
* (not reject)
* `RejectionReason::TransactionProposerTimestampNoLongerValid`

And e.g. would catch us getting the rejection reasons round the wrong way or something :D

**Review by dhedey — APPROVED** [review]
Nice.

Sorry - looks like there's some merge conflicts with https://github.com/radixdlt/radixdlt-scrypto/pull/1943 - should be easy to resolve though I hope 👍

**Pull request merged**